akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala

```diff
@@ -102,6 +103,25 @@ import scala.concurrent.duration._
  */
 @InternalApi private[typed] object ActorContextAdapter {
 
+  private def getLoggerClass(system: ExtendedActorSystem): Class[_] = {
+    // FIXME use stack walker API when we no longer need to support Java 8
+    val trace = Thread.currentThread().getStackTrace
+    var suitableClass: OptionVal[Class[_]] = OptionVal.None
+    var idx = 2 // skip this method and ctx.log right away
+    while (suitableClass.isEmpty && idx < trace.length) {
+      val entry = trace(idx)
+      val name = entry.getClassName
+      if (!name.startsWith("scala.runtime") && !name.startsWith("akka.actor.typed.internal"))
+        system.dynamicAccess.getClassFor[AnyRef](name) match {
+          case Success(clazz) ⇒ suitableClass = OptionVal.Some(clazz)
+          case _              ⇒ // just ignore, we'll hopefully find a class we can load further down
+        }
+      idx += 1
+    }
+
+    suitableClass.getOrElse(classOf[Behavior[_]])
+  }
+
```

Comment:
This isn't water tight, the first log may be called by something else than you'd expect, also, it's kind of expensive. Should it have a setting to toggle it off for performance?

Reply:
could you explain "first log may be called by something else than you'd expect" ?

Reply:
For `MutableBehavior` we have a natural class, but that is probably not used if first log access is from inside the Behaviors.setup block?

Perhaps we should provide a `ActorContext.setLogClass`, and if that is used before first access of `ActorContext.log` it will not fallback to this costly and sometimes surprising way? Perhaps it's not worth having the magic at all?

Reply:
Perhaps we should also think about Lightbend Telemetry and use the same assigned class for metrics grouped by actor class. Perhaps use a more general name such as `setActorClass`?

Reply:
> could you explain "first log may be called by something else than you'd expect" ?

If you are composing some behavior and the (perhaps prewritten/library) thing you compose with call log before your "actual" behavior logic, that will give a surprising class.

In simpler cases the first log access will be what you expect - the scope of the behavior, and the same should be true for MutableBehavior usage (because setup comes from factory method in companion or static factory in behavior class).

I don't like the imperative set-actor-class method because that would add another level of potential confusing ordering of things - if you call it before someone uses the logger but maybe someone does... 

I think the `withLogClass` is better as that requires that you explicitly use the returned logger to get the benefit (just like `withMdc`). I'd potentially agree skipping the magic and just having the `with` variants...

Reply:
So WDYT, should that be the only option:

```Scala
Behaviors.setup { ctx => 
  val log = ctx.log.withLogClass[MyClass]

  log.info("Look a me ma!")
  Behaviors.yadiyadi
}
```

Reply:
I think that is good without surprises, but it would be nice if we can fall back to detecting the class if the call is within an `AbstractBehavior` (and logClass not already defined)